lua/tint.lua

```diff
@@ -414,6 +414,16 @@ __.setup_all = function(skip_config)
 
   setup_namespaces()
   setup_autocmds()
+
+  vim.schedule(function()
```

Comment:
Do we need to wrap this in a `vim.schedule`? Usually if we do, its related to doing something at the wrong time and `vim.schedule` usually just "inappropriately" circumvents that.

Since `setup_all` _is_ called `on_or_after_vimenter`, we should be good to just do this directly as far as I'm aware.

I'm not against merging it as such and don't want to be too much of a pain to get a fix like this merged, but am mostly curious.

Reply:
I have this old vimscript code in a function that is called on VimEnter, that also iterates all buffers and wipes the ones whose files are not readable. My experience was that loop and this loop (tint.nvim's iterate all windows and run on_unfocus on them) were running at the same time and the result was that the current window was tinted after VimEnter.

I'll give you steps to reproduce when I get the chance.

Reply:
> I'll give you steps to reproduce when I get the chance.

The command to run:

```bash
 nvim --clean -S session.vim -u init.vim
```

Try running first with `print_current_winid()` (at the end of `init.vim`). When you're in, run :messages: It outputs 

```
current winid 1000
current winid 1002
```

 `:echo win_getid()` outputs 1002.

Now run `vim.schedule(function() print_current_winid() end)` (at the end of `init.vim`). When you're in, run :messages: It outputs 

```
current winid 1000
```

 `:echo win_getid()` outputs 1002.

The result is that win tint.nvim 1002 is tinted, which is undesirable.

session.vim, a session file which opens 3 windows with these files: `/etc/hosts`, `/etc/hostname`, and `hello` (doesn't exist).

<details>
<summary>Click to expand session.vim</summary>
```vimscript
let SessionLoad = 1
let s:so_save = &g:so | let s:siso_save = &g:siso | setg so=0 siso=0 | setl so=-1 siso=-1
let v:this_session=expand("<sfile>:p")
silent only
silent tabonly
if expand('%') == '' && !&modified && line('$') <= 1 && getline(1) == ''
  let s:wipebuf = bufnr('%')
endif
let s:shortmess_save = &shortmess
if &shortmess =~ 'A'
  set shortmess=aoOA
else
  set shortmess=aoO
endif
badd +1 /etc/hosts
badd +1 /etc/hostname
badd +0 hello
argglobal
%argdel
$argadd /etc/hosts
$argadd /etc/hostname
edit /etc/hosts
let s:save_splitbelow = &splitbelow
let s:save_splitright = &splitright
set splitbelow splitright
wincmd _ | wincmd |
vsplit
1wincmd h
wincmd w
wincmd _ | wincmd |
split
1wincmd k
wincmd w
let &splitbelow = s:save_splitbelow
let &splitright = s:save_splitright
wincmd t
let s:save_winminheight = &winminheight
let s:save_winminwidth = &winminwidth
set winminheight=0
set winheight=1
set winminwidth=0
set winwidth=1
exe 'vert 1resize ' . ((&columns * 119 + 119) / 239)
exe '2resize ' . ((&lines * 23 + 24) / 48)
exe 'vert 2resize ' . ((&columns * 119 + 119) / 239)
exe '3resize ' . ((&lines * 22 + 24) / 48)
exe 'vert 3resize ' . ((&columns * 119 + 119) / 239)
argglobal
setlocal fdm=manual
setlocal fde=0
setlocal fmr={{{,}}}
setlocal fdi=#
setlocal fdl=0
setlocal fml=1
setlocal fdn=20
setlocal fen
silent! normal! zE
let &fdl = &fdl
let s:l = 1 - ((0 * winheight(0) + 23) / 46)
if s:l < 1 | let s:l = 1 | endif
keepjumps exe s:l
normal! zt
keepjumps 1
normal! 0
wincmd w
argglobal
2argu
if bufexists(fnamemodify("hello", ":p")) | buffer hello | else | edit hello | endif
if &buftype ==# 'terminal'
  silent file hello
endif
balt /etc/hostname
setlocal fdm=manual
setlocal fde=0
setlocal fmr={{{,}}}
setlocal fdi=#
setlocal fdl=0
setlocal fml=1
setlocal fdn=20
setlocal fen
silent! normal! zE
let &fdl = &fdl
let s:l = 1 - ((0 * winheight(0) + 11) / 23)
if s:l < 1 | let s:l = 1 | endif
keepjumps exe s:l
normal! zt
keepjumps 1
normal! 0
wincmd w
argglobal
2argu
balt hello
setlocal fdm=manual
setlocal fde=0
setlocal fmr={{{,}}}
setlocal fdi=#
setlocal fdl=0
setlocal fml=1
setlocal fdn=20
setlocal fen
silent! normal! zE
let &fdl = &fdl
let s:l = 1 - ((0 * winheight(0) + 11) / 22)
if s:l < 1 | let s:l = 1 | endif
keepjumps exe s:l
normal! zt
keepjumps 1
normal! 0
wincmd w
2wincmd w
exe 'vert 1resize ' . ((&columns * 119 + 119) / 239)
exe '2resize ' . ((&lines * 23 + 24) / 48)
exe 'vert 2resize ' . ((&columns * 119 + 119) / 239)
exe '3resize ' . ((&lines * 22 + 24) / 48)
exe 'vert 3resize ' . ((&columns * 119 + 119) / 239)
tabnext 1
if exists('s:wipebuf') && len(win_findbuf(s:wipebuf)) == 0 && getbufvar(s:wipebuf, '&buftype') isnot# 'terminal'
  silent exe 'bwipe ' . s:wipebuf
endif
unlet! s:wipebuf
set winheight=1 winwidth=20
let &shortmess = s:shortmess_save
let &winminheight = s:save_winminheight
let &winminwidth = s:save_winminwidth
let s:sx = expand("<sfile>:p:r")."x.vim"
if filereadable(s:sx)
  exe "source " . fnameescape(s:sx)
endif
let &g:so = s:so_save | let &g:siso = s:siso_save
set hlsearch
nohlsearch
doautoall SessionLoadPost
unlet SessionLoad
" vim: set ft=vim :
```
</details>

An init.vim:

<details>
<summary>Click to expand init.vim</summary>
```vimscript
function! s:BWipeNotReadable()
	let buffer_count = 0
    let error_count = 0
    let ei = &eventignore
    set eventignore+=TabClosed
    for buf in getbufinfo()
        let n = buf.bufnr
        if buflisted(n) && !filereadable(bufname(n))
            let command = "bwipe! " . n
            try
                silent exe command
                let buffer_count += 1
            catch
                let error_count += 1
            endtry
        endif
    endfor
    let &eventignore = ei
endfunction

augroup SessionLoad
    autocmd!
    " Wiping empty buffers created by restoring sessions
    autocmd SessionLoadPost * autocmd SessionLoad VimEnter * call s:BWipeNotReadable()
augroup END

" autocmd VimEnter * call s:BWipeNotReadable()

lua << EOF
local function iterate_all_windows(func)
  for _, tabpage in ipairs(vim.api.nvim_list_tabpages()) do
    if vim.api.nvim_tabpage_is_valid(tabpage) then
      for _, winid in ipairs(vim.api.nvim_tabpage_list_wins(tabpage)) do
        if vim.api.nvim_win_is_valid(winid) then
          func(winid, tabpage)
        end
      end
    end
  end
end

local function print_current_winid()
  iterate_all_windows(function(winid, _)
    if vim.api.nvim_get_current_win() ~= winid then
      vim.api.nvim_win_call(winid, function()
        print("current winid "..vim.api.nvim_get_current_win())
      end)
    end
  end)
end

vim.api.nvim_create_autocmd("VimEnter", {
  callback = function()
    print_current_winid()
    -- vim.schedule(function() print_current_winid() end)
  end,
})
EOF
```
</details>